<a href="https://colab.research.google.com/github/Manuelamejiah/Business-intelligence-/blob/main/%C3%81rboles_de_decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Obj. implementar modelos de arboles de decisión
Entradas: EDAD, INGRESOS, EGRESOS Y MONTO
Salidas: preapr
se pronosticara la preapr a solicitantes que no lo tienen

Cargamos las librerias de trabajo

In [ ]:
import pandas as pd
import numpy as np

#librerias arbol
from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import confusion_matrix

#conectar al drive -> vamos a tener los archivos disponibles
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
nxl='/content/drive/MyDrive/ANALÍTICA DE NEGOCIOS /6. SolicitantesCrédito(USD) (3).xlsx' #NXL SIGNIFICA NAME EXCEL
XDB=pd.read_excel(nxl, sheet_name=1) #base AAA
XDB=XDB.dropna()
XDB.head(100)





Unnamed: 0  Edad    Sexo Estado civil  Hijos  Perscargo  Nivel Renta  \
0            1    38      NR       Casado      1          0            0   
1            2    51      NR     Separado      6          1            1   
2            3    37   Mujer       Casado      2          0            0   
3            4    29      NR     Separado      1          0            0   
4            5    42   Mujer       Casado      3          4            4   
..         ...   ...     ...          ...    ...        ...          ...   
95         130    50   Mujer      Soltero      2          1            1   
96         131    62  Hombre      Soltero      1          1            1   
97         132    54      NR       Casado      0          1            1   
98         136    54   Mujer      Soltero      2          0            0   
99         137    47   Mujer     Separado      1          1            1   

    Estrato Nivel de estudios  Tipo de contrato  ...  Plazo  Cuota (COP)  \
0         3         Bachiller           Ninguno  ...     48    18.375890   
1         4           Técnico           Ninguno  ...     36     5.062492   
2         3         Tecnólogo        Indefinido  ...     36    10.287688   
3         2         Tecnólogo           Ninguno  ...     24    15.458878   
4         2   Especializacion           Ninguno  ...     24    24.995180   
..      ...               ...               ...  ...    ...          ...   
95        4       Profesional  Labor Contratada  ...     36    26.659418   
96        3         Bachiller        Indefinido  ...     36     1.512706   
97        4       Profesional        Indefinido  ...      3    72.959992   
98        2           Técnico        Indefinido  ...     36     1.751685   
99        5           Ninguno           Ninguno  ...     48    51.201558   

              ConoCliente              Garantia            RespPat  \
0      Buenas Referencias          Sin Codeudor     Sin Patrimonio   
1      Buenas Referencias          Sin Codeudor  Prop.sin Hipoteca   
2      Buenas Referencias      Con Codeudor(es)     Sin Patrimonio   
3      Buenas Referencias      Con Codeudor(es)     Sin Patrimonio   
4      Buenas Referencias      Con Codeudor(es)     Sin Patrimonio   
..                    ...                   ...                ...   
95  Buena Experiencia CFB  Libranza Establecida  Prop.sin Hipoteca   
96  Buena Experiencia CFB          Sin Codeudor  Prop.sin Hipoteca   
97  Buena Experiencia CFB          Sin Codeudor     Sin Patrimonio   
98  Buena Experiencia CFB          Sin Codeudor     Sin Patrimonio   
99  Buena Experiencia CFB          Sin Codeudor  Prop.con Hipoteca   

                    MoraCome  PreApr  Prob.Default (PD)       Score       LGD  
0   Con Experiencia sin Mora       1           0.041659  582.478943  0.063702  
1   Con Experiencia sin Mora       1           0.234951  479.839966  0.465636  
2   Con Experiencia sin Mora       0           0.170330  417.653107  0.290071  
3   Con Experiencia sin Mora       1           0.201659  316.791107  0.278389  
4   Con Experiencia sin Mora       1           0.140929  441.809570  0.203359  
..                       ...     ...                ...         ...       ...  
95  Con Experiencia sin Mora       0           0.063148  594.865723  0.120084  
96  Con Experiencia sin Mora       0           0.303231  406.213745  0.558065  
97  Con Experiencia sin Mora       0           0.116497  436.862579  0.216330  
98  Con Experiencia sin Mora       0           0.316390  354.029877  0.520874  
99  Con Experiencia sin Mora       1           0.006421  866.446594  0.009203  

[100 rows x 36 columns]

**Descripción de los datos**


De acuerdo con la bd de solicitantes de crédito, esta posee un total de 5842 datos totales divididos en 2959 prenegados y 2883 preaprobados. Cada uno de los solicitantes estará descrito en función de 4 variables,
* Edad: Indica el número de años de la persona
* Ingresos: Indica la cantidad de ingresos totales registrados en un mes por el solicitante  
* Egresos: Indica los gastos en los que incurre en el mes una persona pra cubirir sus necesidades básicas
* Monto (EAD): Es la cantidad solicitada de una persona a la fintech
* PreApr: Indica la decisión del banco frente al otorgamiento del crédito

In [ ]:
#seleccionar variables
XD=XDB[['Edad','Ingresos','Egresos','Monto (EAD)']]
YD=XDB[['PreApr']]

2. se procede con la implememntación del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,YD)

#se procede con la gráfica
from pydotplus import graph_from_dot_data #me crea la gráfica
from sklearn.tree import export_graphviz #me exporta a png, jpg, tiff

ve= ['Edad','Ingresos','Egresos','Monto (EAD)']
dot_data=export_graphviz(mar,feature_names=ve)
graph_from_dot_data(dot_data).write_png('2. ArbolDeDecisión_M.png')




True

3. se procede con el pronóstico de los solicitantes que no poseen preaprobación

In [ ]:
XDB0=pd.read_excel(nxl, sheet_name=0)
XDB0.head(100)
#vamos a clasificar los que están en la hoja 1 con los solicitantes de crédito de la hoja 2
#Se cre al bd a pronosticar
XD0=XDB0[['Edad','Ingresos','Egresos','Monto (EAD)']]
YD0=XDB0[['PreApr']]

#determinar filas de solicitantes que no poseen preaprb
filas=np.where(np.isnan(YD0))[0]
XD0=np.array(XD0)
XD0=XD0[filas]

#le calculaos la preaprobación a los solicitantes
y_pred=mar.predict(XD0)
print(y_pred)

#se crea el informe en excel
df=pd.DataFrame(np.column_stack((filas, XD0,y_pred)))
#filas donde estan los solicitantes que no poseen preapr, XD0 los datos de esas eprsonas, y_pred la decisión
df.columns=['Solicitante', 'Edad','Ingresos','Egresos','Monto (EAD)','PreApr']
df.to_excel("Evaluación créditos.xlsx")

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


[0 0 1 ... 0 1 0]


4. Evaluar el comportamiento del modelo

In [ ]:
ydp=mar.predict(XD) #pronóstico sobre la bd original
cm=confusion_matrix(YD,ydp)
print("La matriz de confusión es:\n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]


#Métricas de desempeño
EX=(VP+VN)/(VP+VN+FP+FN) #Exactitud
print('La exactitud del modelo es:\n',EX)

TE=(FN+FP)/(VP+VN+FP+FN) #Tasa de error:
print('La tasa de error del modelo es:\n',TE)

SE=VP/(VP+FN) #Sensibilidad
print('La sensibilidad del modelo es:\n',SE)

SP=VN/(VN+FP) #Especificidad
print('La especificidad del modelo es:\n',SP)

PR=VP/(VP+FP) #Precisión
print('La precisión del modelo es:\n',PR)

PRNeng=VN/(VN+FN) #Precisión Negativa
print('La precisión negativa del modelo es:\n',PRNeng)

La matriz de confusión es:
 [[2301  658]
 [ 644 2239]]
La exactitud del modelo es:
 0.7771311194796303
La tasa de error del modelo es:
 0.22286888052036974
La sensibilidad del modelo es:
 0.7766215747485259
La especificidad del modelo es:
 0.7776275768840825
La precisión del modelo es:
 0.772868484639282
La precisión negativa del modelo es:
 0.7813242784380305


**Análisis de resultados**

La Matriz de confusión, muestra una predominancia por los valores verdaderos negativos (2301) y los valores verdaderos positivos (2239). Lo anterior hizo que el modelo alcanzara valores por encima del 75% en promedio para los valores de exactitud, sensibilidad, procisión y especificidad, lo que muestra claramente el buen funcionamiento del modelo frente al pronóstico de la preabrobación. Con respecto al pronóstico de nuevos solicitantes de crédito, la bd mostró un total de 2533 solicitantes, sin decisión de crédito. Se destacan un total de (#) de solicitudes aprobados y (#neg) prenegados. Se destaca que el número de (preneg o preapr) estuvo por encima del número de (preneg o preapr)

De acuerdo los índices, el modelo muestra un buen comportamiento frente a la especificidad, o clasificación de prenegados, a diferencia de la sensibilidad, la cual estuvo por encima del (75%), lo que indica su comportamiento frenet al modelo ma